File: algorithms/math/catalan-numbers/scala/CatalanNumbers.scala
object CatalanNumbers {

  val MOD: Long = 1000000007L

  def modPow(base: Long, exp: Long, mod: Long): Long = {
    var result = 1L
    var b = base % mod
    var e = exp
    while (e > 0) {
      if (e % 2 == 1) result = result * b % mod
      e /= 2
      b = b * b % mod
    }
    result
  }

  def modInv(a: Long, mod: Long): Long = modPow(a, mod - 2, mod)

  def catalanNumbers(n: Int): Int = {
    var result = 1L
    for (i <- 1 to n) {
      result = result * (2L * (2 * i - 1)) % MOD
      result = result * modInv(i + 1, MOD) % MOD
    }
    result.toInt
  }
}

File: algorithms/math/catalan-numbers/c/catalan_numbers.c
#include "catalan_numbers.h"

static const long long MOD = 1000000007;

static long long mod_pow(long long base, long long exp, long long mod) {
    long long result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp % 2 == 1) result = result * base % mod;
        exp /= 2;
        base = base * base % mod;
    }
    return result;
}

static long long mod_inv(long long a, long long mod) {
    return mod_pow(a, mod - 2, mod);
}

int catalan_numbers(int n) {
    long long result = 1;
    for (int i = 1; i <= n; i++) {
        result = result * (2LL * (2 * i - 1)) % MOD;
        result = result * mod_inv(i + 1, MOD) % MOD;
    }
    return (int)result;
}
